File: Algorithms/Challenges/Arrays/OddCountElements/Python/OddCountElements.py
"""
Description
Return all integers that appear an odd number of times in an array

Runtime: O(n)
Space: O(n)

Example:

Input: int_array
5 1 2 4 4 3 4 1 5

Output: int_array
2 3 4
"""

def stringAsIntArr(str):
    elems = str.split(" ")
    arr = []
    for elem in elems:
       arr.append(int(elem))
    return arr;

def oddTimeElems(arr):
    result = set()
    dict = {}
    for i in arr:
        if i not in dict:
            dict[i] = 1
            result.add(i)
        else:
            dict[i] = dict[i] + 1
            if dict[i] % 2 == 0:
                result.remove(i)
            else:
                result.add(i)
    return list(result)

def solve(input):
    arr = stringAsIntArr(input)
    return " ".join(map(str, oddTimeElems(arr)))

input_file = open("io/tests.txt", "r");
output_file = open("io/solution.txt", "w");
debug = True

test = input_file.readline()
while test != '':
    test = test.rstrip("\r\n")
    solution = solve(test)
    if debug: solution +=  ": '%s'" % test
    output_file.write(solution + "\n")
    test = input_file.readline()

input_file.close();
output_file.close();

File: Algorithms/Challenges/Arrays/OddCountElements/Java/OddCountElements.java
import java.util.*;
import java.io.*;

/*
Description
Return all integers that appear an odd number of times in an array

Runtime: O(n)
Space: O(n)

Example:

Input: int_array
[]

Output: [format]
[]
*/

class OddCountElements {

  static int[] stringAsIntArr(String str) {
    String[] elems = str.split(" ");
    int[] arr = new int[elems.length];
    for(int i = 0; i < elems.length; i++) {
       arr[i] = Integer.parseInt(elems[i]);
    }
    return arr;
  }

  static String intArrAsString(int[] arr) {
    StringBuilder sb = new StringBuilder();
    for(int i = 0; i < arr.length; i++) {
       sb.append(arr[i]);
       if (i != arr.length - 1)
        sb.append(" ");
    }
    return sb.toString();
  }

  public static int[] convertIntegers(TreeSet<Integer> integers){
    int[] ret = new int[integers.size()];
    Iterator<Integer> iterator = integers.iterator();
    for (int i = 0; i < ret.length; i++) ret[i] = iterator.next().intValue();
    return ret;
  }

  public static int[] oddTimeElems(int[] arr) {
    TreeSet<Integer> result = new TreeSet<Integer>();
    HashMap<Integer, Integer> hm = new HashMap<Integer, Integer>();
    for (int i : arr) {
      if (hm.get(i) == null) {
        hm.put(i, 1);
        result.add(i);
      } else {
        hm.put(i, hm.get(i) + 1);
        if (hm.get(i) % 2 == 0)
          result.remove(i);
        else
          result.add(i);
      }
    }
    int[] answer = convertIntegers(result);
    return answer;
  }

  static String solve(String str) {
    int[] arr = stringAsIntArr(str);
    return intArrAsString(oddTimeElems(arr));
  }

  public static void main(String[] args) throws IOException, FileNotFoundException {
    BufferedReader br = new BufferedReader(new FileReader("io/tests.txt"));
    BufferedWriter bw = new BufferedWriter(new FileWriter("io/solution.txt"));
    boolean debug = true;

    OddCountElements classInstance = new OddCountElements(); // If non-static methods

    String test;
    while ( (test = br.readLine()) != null ) {
      String solution = String.valueOf(solve(test));
      if (debug) solution +=  ": '" + test + "'";
      bw.write(solution);
      bw.newLine();
    }

    br.close();
    bw.close();
  }
}
